File: internal/scheduler/phase2.go
package scheduler

import (
	"fmt"
	"math/rand"
	"sort"

	"github.com/pseudomuto/skedge/internal/config"
)

// phase2 fills Block.Teacher and Block.Room for every slot in schedule using
// backtracking search. Targets are ordered globally most-constrained-first so
// that hard constraints (e.g. single-teacher subjects) are detected and fail-fast
// early, rather than after exhausting large portions of the search space.
func phase2(cfg *config.Config, schedule []*Class, rng *rand.Rand) error {
	type slotKey struct {
		day   string
		block int
	}
	type authKey struct {
		teacher string
		class   string
	}

	teacherInSlot := make(map[slotKey]map[string]bool)
	roomInSlot := make(map[slotKey]map[string]bool)
	teacherTotal := make(map[string]int)

	authorized := make(map[authKey]map[string]struct{})
	for _, t := range cfg.Teachers {
		for _, ts := range t.Subjects {
			k := authKey{t.Name, ts.Class}
			if authorized[k] == nil {
				authorized[k] = make(map[string]struct{})
			}
			for _, s := range ts.Subjects {
				authorized[k][s] = struct{}{}
			}
		}
	}

	eligible := func(t config.Teacher, className, subject string, sk slotKey) bool {
		subjects, ok := authorized[authKey{t.Name, className}]
		if !ok {
			return false
		}
		if _, ok := subjects[subject]; !ok {
			return false
		}
		if teacherInSlot[sk][t.Name] {
			return false
		}
		if roomInSlot[sk][t.Room] {
			return false
		}
		return teacherTotal[t.Name] < maxTeacherBlocksPerWeek
	}

	type target struct {
		className    string
		cohort       *Cohort
		dayIdx       int
		blockIdx     int
		sk           slotKey
		initialCount int
	}

	// Build targets and compute initial eligible counts before any assignments.
	// Targets are then sorted most-constrained-first so single-teacher subjects
	// (e.g. Music → only Gumaste) are processed first and conflicts surface immediately.
	var targets []target
	for _, class := range schedule {
		if len(class.Cohorts) == 0 {
			continue
		}
		numDays := len(class.Cohorts[0].Schedule)
		numBlocks := len(class.Cohorts[0].Schedule[0].Blocks)
		for di := range numDays {
			dayName := class.Cohorts[0].Schedule[di].Day
			for bi := range numBlocks {
				sk := slotKey{dayName, bi}
				if teacherInSlot[sk] == nil {
					teacherInSlot[sk] = make(map[string]bool)
				}
				if roomInSlot[sk] == nil {
					roomInSlot[sk] = make(map[string]bool)
				}
				for ci := range class.Cohorts {
					subject := class.Cohorts[ci].Schedule[di].Blocks[bi].Name
					count := 0
					for _, t := range cfg.Teachers {
						if eligible(t, class.Name, subject, sk) {
							count++
						}
					}
					targets = append(targets, target{
						className:    class.Name,
						cohort:       &class.Cohorts[ci],
						dayIdx:       di,
						blockIdx:     bi,
						sk:           sk,
						initialCount: count,
					})
				}
			}
		}
	}

	sort.SliceStable(targets, func(i, j int) bool {
		return targets[i].initialCount < targets[j].initialCount
	})

	// Shuffle teachers so different retry attempts explore different orderings.
	teachers := make([]config.Teacher, len(cfg.Teachers))
	copy(teachers, cfg.Teachers)
	rng.Shuffle(len(teachers), func(i, j int) { teachers[i], teachers[j] = teachers[j], teachers[i] })

	n := len(targets)
	m := len(teachers)

	// Precompute slot indices so slot membership can be checked via arrays instead
	// of map lookups. This makes each MRV scan ~10× faster.
	var slots []slotKey
	slotIdx := make(map[slotKey]int)
	for _, tgt := range targets {
		if _, ok := slotIdx[tgt.sk]; !ok {
			slotIdx[tgt.sk] = len(slots)
			slots = append(slots, tgt.sk)
		}
	}
	slotForTarget := make([]int, n)
	for fi, tgt := range targets {
		slotForTarget[fi] = slotIdx[tgt.sk]
	}

	// Precompute room indices.
	roomIdxMap := make(map[string]int)
	for _, t := range teachers {
		if _, ok := roomIdxMap[t.Room]; !ok {
			roomIdxMap[t.Room] = len(roomIdxMap)
		}
	}
	roomForTeacher := make([]int, m)
	for ti, t := range teachers {
		roomForTeacher[ti] = roomIdxMap[t.Room]
	}

	// canTeach[fi][ti] = teacher[ti] is authorized for the subject of target[fi].
	canTeach := make([][]bool, n)
	for fi, tgt := range targets {
		canTeach[fi] = make([]bool, m)
		blockName := tgt.cohort.Schedule[tgt.dayIdx].Blocks[tgt.blockIdx].Name
		for ti, t := range teachers {
			if eligible(t, tgt.className, blockName, tgt.sk) {
				canTeach[fi][ti] = true
			}
		}
	}

	// Array-based state: avoids per-call map lookups inside the hot MRV loop.
	inSlot := make([][]bool, m)     // inSlot[ti][si]
	inRoom := make([][]bool, len(roomIdxMap)) // inRoom[ri][si]
	for ti := range teachers {
		inSlot[ti] = make([]bool, len(slots))
	}
	for ri := range inRoom {
		inRoom[ri] = make([]bool, len(slots))
	}
	remaining := make([]int, m)
	for ti := range remaining {
		remaining[ti] = maxTeacherBlocksPerWeek
	}

	fastEligible := func(fi, ti int) bool {
		si := slotForTarget[fi]
		return canTeach[fi][ti] && !inSlot[ti][si] && !inRoom[roomForTeacher[ti]][si] && remaining[ti] > 0
	}

	// MRV backtracking with a node budget. We pick the unassigned target with the
	// fewest currently eligible teachers (minimum remaining values). A target hitting
	// 0 eligible teachers is detected immediately (forward check). The budget caps
	// work on phase1 outputs that are globally infeasible in ways only detectable
	// late in the search; those attempts are abandoned so the outer retry loop can
	// try a fresh phase1 shuffle. Feasible configs are solved in O(n) MRV steps.
	const nodeBudget = 5_000
	assigned := make([]bool, n)
	nodes := 0

	var bt func() bool
	bt = func() bool {
		if nodes++; nodes > nodeBudget {
			return false
		}
		bestIdx := -1
		bestCount := m + 1
		for fi := range targets {
			if assigned[fi] {
				continue
			}
			count := 0
			for ti := range teachers {
				if fastEligible(fi, ti) {
					count++
				}
			}
			if count == 0 {
				return false
			}
			if count < bestCount {
				bestCount = count
				bestIdx = fi
				if bestCount == 1 {
					break // can't do better; assign immediately
				}
			}
		}

		if bestIdx == -1 {
			return true
		}

		assigned[bestIdx] = true
		tgt := targets[bestIdx]
		b := &tgt.cohort.Schedule[tgt.dayIdx].Blocks[tgt.blockIdx]
		si := slotForTarget[bestIdx]

		for ti, t := range teachers {
			if !fastEligible(bestIdx, ti) {
				continue
			}
			b.Teacher = t.Name
			b.Room = t.Room
			ri := roomForTeacher[ti]
			inSlot[ti][si] = true
			inRoom[ri][si] = true
			remaining[ti]--

			if bt() {
				return true
			}

			b.Teacher = ""
			b.Room = ""
			inSlot[ti][si] = false
			inRoom[ri][si] = false
			remaining[ti]++
		}

		assigned[bestIdx] = false
		return false
	}

	if !bt() {
		return fmt.Errorf("no valid teacher assignment found")
	}
	return nil
}

File: web/src/scheduler/phase2.ts
import type { Config, Teacher } from '../types/config'
import type { ScheduleClass } from '../types/schedule'
import { shuffle } from './utils'
import { MAX_TEACHER_BLOCKS_PER_WEEK } from './validate'

const NODE_BUDGET = 5_000

function slotKey(day: string, blockIdx: number): string {
  return `${day}:${blockIdx}`
}

export function phase2(cfg: Config, schedule: ScheduleClass[], rng: () => number): Error | null {
  // Build authorization map: "teacher:class" -> Set<subject>
  const authorized = new Map<string, Set<string>>()
  for (const t of cfg.teachers) {
    for (const ts of t.subjects) {
      const k = `${t.name}:${ts.class}`
      if (!authorized.has(k)) authorized.set(k, new Set())
      for (const s of ts.subjects) authorized.get(k)!.add(s)
    }
  }

  // Precompute slot keys and indices
  const slotIdxMap = new Map<string, number>()
  const slots: string[] = []

  function getSlotIdx(sk: string): number {
    let idx = slotIdxMap.get(sk)
    if (idx === undefined) {
      idx = slots.length
      slotIdxMap.set(sk, idx)
      slots.push(sk)
    }
    return idx
  }

  type Target = {
    className: string
    cohortIdx: number
    classIdx: number
    dayIdx: number
    blockIdx: number
    sk: string
    initialCount: number
  }

  // First pass: collect all slots we need
  for (const cls of schedule) {
    if (cls.cohorts.length === 0) continue
    const numDays = cls.cohorts[0].schedule.length
    const numBlocks = cls.cohorts[0].schedule[0].blocks.length
    for (let di = 0; di < numDays; di++) {
      const dayName = cls.cohorts[0].schedule[di].day
      for (let bi = 0; bi < numBlocks; bi++) {
        getSlotIdx(slotKey(dayName, bi))
      }
    }
  }

  const numSlots = slots.length

  // Shuffle teachers for randomization
  const teachers: Teacher[] = [...cfg.teachers]
  shuffle(teachers, rng)
  const m = teachers.length

  // Room index map
  const roomIdxMap = new Map<string, number>()
  for (const t of teachers) {
    if (!roomIdxMap.has(t.room)) roomIdxMap.set(t.room, roomIdxMap.size)
  }
  const roomForTeacher: number[] = teachers.map((t) => roomIdxMap.get(t.room)!)
  const numRooms = roomIdxMap.size

  // Array-based state: inSlot[ti][si], inRoom[ri][si], remaining[ti]
  const inSlot: boolean[][] = Array.from({ length: m }, () => new Array<boolean>(numSlots).fill(false))
  const inRoom: boolean[][] = Array.from({ length: numRooms }, () => new Array<boolean>(numSlots).fill(false))
  const remaining: number[] = new Array<number>(m).fill(MAX_TEACHER_BLOCKS_PER_WEEK)

  // Build eligibility map for initial count (before any assignments)
  // We need to check authorization only (not slot conflicts, since nothing assigned yet)
  function isAuthorized(ti: number, className: string, subject: string): boolean {
    const k = `${teachers[ti].name}:${className}`
    return authorized.get(k)?.has(subject) ?? false
  }

  // Build targets
  const targets: Target[] = []
  for (let classIdx = 0; classIdx < schedule.length; classIdx++) {
    const cls = schedule[classIdx]
    if (cls.cohorts.length === 0) continue
    const numDays = cls.cohorts[0].schedule.length
    const numBlocks = cls.cohorts[0].schedule[0].blocks.length
    for (let di = 0; di < numDays; di++) {
      const dayName = cls.cohorts[0].schedule[di].day
      for (let bi = 0; bi < numBlocks; bi++) {
        const sk = slotKey(dayName, bi)
        for (let cohortIdx = 0; cohortIdx < cls.cohorts.length; cohortIdx++) {
          const subject = cls.cohorts[cohortIdx].schedule[di].blocks[bi].name
          let count = 0
          for (let ti = 0; ti < m; ti++) {
            if (isAuthorized(ti, cls.name, subject)) count++
          }
          targets.push({
            className: cls.name,
            cohortIdx,
            classIdx,
            dayIdx: di,
            blockIdx: bi,
            sk,
            initialCount: count,
          })
        }
      }
    }
  }

  // Sort by initial eligible count ascending (MRV)
  targets.sort((a, b) => a.initialCount - b.initialCount)

  const n = targets.length

  // Precompute canTeach[fi][ti]: teacher ti can teach subject for target fi
  // This is authorization-only (slot conflicts handled dynamically)
  const canTeach: boolean[][] = targets.map((tgt) => {
    const subject = schedule[tgt.classIdx].cohorts[tgt.cohortIdx].schedule[tgt.dayIdx].blocks[tgt.blockIdx].name
    return teachers.map((_, ti) => isAuthorized(ti, tgt.className, subject))
  })

  // Precompute slot index for each target
  const slotForTarget: number[] = targets.map((tgt) => slotIdxMap.get(tgt.sk)!)

  function fastEligible(fi: number, ti: number): boolean {
    const si = slotForTarget[fi]
    return canTeach[fi][ti] && !inSlot[ti][si] && !inRoom[roomForTeacher[ti]][si] && remaining[ti] > 0
  }

  const assigned: boolean[] = new Array<boolean>(n).fill(false)
  let nodes = 0

  function bt(): boolean {
    if (++nodes > NODE_BUDGET) return false

    // MRV: find unassigned target with fewest eligible teachers
    let bestIdx = -1
    let bestCount = m + 1
    for (let fi = 0; fi < n; fi++) {
      if (assigned[fi]) continue
      let count = 0
      for (let ti = 0; ti < m; ti++) {
        if (fastEligible(fi, ti)) count++
      }
      if (count === 0) return false
      if (count < bestCount) {
        bestCount = count
        bestIdx = fi
        if (bestCount === 1) break
      }
    }
    if (bestIdx === -1) return true

    assigned[bestIdx] = true
    const tgt = targets[bestIdx]
    const b = schedule[tgt.classIdx].cohorts[tgt.cohortIdx].schedule[tgt.dayIdx].blocks[tgt.blockIdx]
    const si = slotForTarget[bestIdx]

    for (let ti = 0; ti < m; ti++) {
      if (!fastEligible(bestIdx, ti)) continue
      const t = teachers[ti]
      b.teacher = t.name
      b.room = t.room
      const ri = roomForTeacher[ti]
      inSlot[ti][si] = true
      inRoom[ri][si] = true
      remaining[ti]--

      if (bt()) return true

      b.teacher = ''
      b.room = ''
      inSlot[ti][si] = false
      inRoom[ri][si] = false
      remaining[ti]++
    }

    assigned[bestIdx] = false
    return false
  }

  if (!bt()) return new Error('no valid teacher assignment found')
  return null
}
